com.microsoft.mrtk.uxcore/Keyboard/NonNative/NonNativeKeyboard.cs

```diff
@@ -0,0 +1,732 @@
+                        break;
+
+                    case Function.Previous:
+                        {
+                            MoveCaretLeft();
+                            break;
+                        }
+
+                    case Function.Next:
+                        {
+                            MoveCaretRight();
+                            break;
+                        }
+
+                    case Function.Close:
+                        Close();
+                        break;
+
+                    case Function.Shift:
+                        Shift(!IsShifted);
+                        break;
+
+                    case Function.CapsLock:
+                        CapsLock(!IsCapsLocked);
+                        break;
+
+                    case Function.Space:
+                        Space();
+                        break;
+
+                    case Function.Backspace:
+                        Backspace();
+                        break;
+
+                    case Function.UNDEFINED:
+                        Debug.LogErrorFormat("The {0} key on this keyboard hasn't been assigned a function.", functionKey.name);
+                        break;
+                }
+            }
+            else
+            {
+                Debug.LogError("You must set an input field for the NonNativeKeyboard.");
+            }
+        }
+
+        /// <summary>
+        /// Delete the last character.
+        /// </summary>
+        public void Backspace()
+        {
+            // check if text is selected
+            if (InputField.selectionFocusPosition != InputField.caretPosition || InputField.selectionAnchorPosition != InputField.caretPosition)
+            {
+                if (InputField.selectionAnchorPosition > InputField.selectionFocusPosition) // right to left
+                {
+                    InputField.text = InputField.text.Substring(0, InputField.selectionFocusPosition) + InputField.text.Substring(InputField.selectionAnchorPosition);
+                    InputField.caretPosition = InputField.selectionFocusPosition;
+                }
+                else // left to right
+                {
+                    InputField.text = InputField.text.Substring(0, InputField.selectionAnchorPosition) + InputField.text.Substring(InputField.selectionFocusPosition);
+                    InputField.caretPosition = InputField.selectionAnchorPosition;
+                }
+
+                m_CaretPosition = InputField.caretPosition;
+                InputField.selectionAnchorPosition = m_CaretPosition;
+                InputField.selectionFocusPosition = m_CaretPosition;
+            }
+            else
+            {
+                m_CaretPosition = InputField.caretPosition;
+
+                if (m_CaretPosition > 0)
+                {
+                    --m_CaretPosition;
+                    InputField.text = InputField.text.Remove(m_CaretPosition, 1);
```

Comment:
I think having a "Text" property that wraps InputField.text would be useful.

This way callers to NonNativeKeyboards don't need to understand how text is stored.

Can you add

public string Text 
{
     get=> InputField == null ? null : InputField.Text;
     set => if (InputField != null) InputField.Text = value;
}

Reply:
And then use the Text property within this class